decepticonlp/transforms/perturbations.py

```diff
@@ -0,0 +1,317 @@
+    """
+
+    def apply(self, word: str, **kwargs):
+        """
+            shifts a character by one keyboard space:
+            one space up, down, left or right
+            each word is typofied with some probability 'p':
+            1. (p*100) percent of character will become typos
+            keyboard is defined as:
+            qwertyuiop
+            asdfghjkl
+             zxcvbnm
+            word = "Noise"
+            print(typo('Noise',0.1))
+            Noide
+            :param word : word to be shuffled
+            :param probability: probability of a typo
+            returns typofied word
+            """
+
+        if kwargs.get("ignore", self.get_ignore_default_value()) and (" " in word):
+            return word
+
+        assert " " not in word, self.get_string_not_a_word_error_msg()
+
+        # convert word to list (string is immutable)
+        word = list(word)
+
+        num_chars_to_shift = math.ceil(len(word) * kwargs.get("probability", 0.1))
+
+        # checking for capitalizations
+        capitalization = [False] * len(word)
+
+        # convert to lowercase and record capitalization
+        for i in range(len(word)):
+            capitalization[i] = word[i].isupper()
+            word[i] = word[i].lower()
+
+        # list of characters to be switched
+        positions_to_shift = []
+        for i in range(num_chars_to_shift):
+            positions_to_shift.append(random.randint(0, len(word) - 1))
+
+        # defining a dictionary of keys located close to each character
+        keys_in_proximity = {
+            "a": ["q", "w", "s", "x", "z"],
+            "b": ["v", "g", "h", "n"],
+            "c": ["x", "d", "f", "v"],
+            "d": ["s", "e", "r", "f", "c", "x"],
+            "e": ["w", "s", "d", "r"],
+            "f": ["d", "r", "t", "g", "v", "c"],
+            "g": ["f", "t", "y", "h", "b", "v"],
+            "h": ["g", "y", "u", "j", "n", "b"],
+            "i": ["u", "j", "k", "o"],
+            "j": ["h", "u", "i", "k", "n", "m"],
+            "k": ["j", "i", "o", "l", "m"],
+            "l": ["k", "o", "p"],
+            "m": ["n", "j", "k", "l"],
+            "n": ["b", "h", "j", "m"],
+            "o": ["i", "k", "l", "p"],
+            "p": ["o", "l"],
+            "q": ["w", "a", "s"],
+            "r": ["e", "d", "f", "t"],
+            "s": ["w", "e", "d", "x", "z", "a"],
+            "t": ["r", "f", "g", "y"],
+            "u": ["y", "h", "j", "i"],
+            "v": ["c", "f", "g", "v", "b"],
+            "w": ["q", "a", "s", "e"],
+            "x": ["z", "s", "d", "c"],
+            "y": ["t", "g", "h", "u"],
+            "z": ["a", "s", "x"],
+        }
+
+        # insert typo
+        for pos in positions_to_shift:
+            # no typo insertion for special characters
+            try:
+                typo_list = keys_in_proximity[word[pos]]
+                word[pos] = random.choice(typo_list)
+            except:
+                break
```

Comment:
Maybe a better way to do this would be to check `word[pos]` in `keys_in_proximity.keys()`? As having a try, except statement is not the safest way. 

If not, maybe we could `except KeyError` and then print out a warning that keys in proximity do not exist for `char {}`

Reply:
You'd probably know better @rohts-patil . So let me know what you think

Reply:
Actually I just refactored the code, this function was already present. I just wrapped it into a class. 
And yes, sonar is reporting this as a critical code smell and we need to fix it. Can we contact the original author?

Reply:
Ok sure. Sounds good.

Reply:
Yes I guess @parantak Authored it, we can implement this

Reply:
@Sharad24 @someshsingh22  The right way is to expect KeyError, log it as warning/error with a suitable user-friendly message.

Reply:
@rohts-patil I am taking this then

Reply:
I understand your concern @someshsingh22. I did not say that the function should throw an error, we should just log it anyhow. If you think that the above scenario is too obvious and will occur many times, we can log it in debug mode, so that the performance does not take a hit even though there are errors. If we check keys each time which will be a linear search I assume it will impact performance for huge sentences/text. Getting value from a dictionary is O(1).
Or simply we can just use the default value of dictionary. 
`keys_in_proximity[word[pos],[word[pos]]] ` So this snippet of code will retrun the character itself if the key is not present.

Reply:
@rohts-patil No, I didn't mean performance issues at all, in fact, that is what I have nearly done in the latest PR. I meant by logging errors did you mean the logging library that creates a .log file or just messages in print? because print would unnecessarily clutter and we haven't used the logging library yet